### ISSUE TITLE
Allow filtering which source sets detekt checks

### ISSUE THREAD

**Issue by matejdro** [detekt/detekt#7927]
Currently Detekt just takes Kotlin tasks and adds all of its sources to its check source set: https://github.com/detekt/detekt/blob/03775f9ff49a449e683a141b3864b40b629f4d04/detekt-gradle-plugin/src/main/kotlin/dev/detekt/gradle/plugin/internal/SharedTasks.kt#L26

Problem is that this includes both actual written code sources from `/src` directory and generated sources (from `/build` directory`). Inspecting generated code is wasteful at best and breaking at worst, because detekt reports issues that developer cannot fix since they are in generated code.

`exclude` option with pattern is not helpful here, because excludes in gradle only works relative to each source set. So you cannot exclude whole source sets that way or even detect which source set is in `build` directory. Exclude with lambda works (as described in the comments of the #5611), but this executes lambda for every single file in your source sets, which can possibly slow down large projects.

My solution for this is using `compilation.kotlinSourceSets.kotlin.srcDirs` instead of `compilation.compileTaskProvider.sources` (https://github.com/JetBrains/kotlin/blob/cd198db772530f4114f1d1329cef27bd622d89ac/libraries/tools/kotlin-gradle-plugin-api/src/common/kotlin/org/jetbrains/kotlin/gradle/plugin/KotlinCompilation.kt#L114). With this, we can actually apply filtering (for example `compilation.kotlinSourceSets.kotlin.srcDirs.filter { !it.relativeTo(projectDir).startsWith("build")}`  to exclude all source sets from the build directory)

With this system, detekt can properly ignore unwanted kotlin source sets. This could be used in two ways:

1. Exclude all source sets originating from `build` directory by default
2. Allow user to set its own filter. Maybe by adding a hook to `DetektExtension` that gets called every time tasks are added?

If accepted, I would be open to creating a PR for this

This would solve #7644, #5611 and provide a better way of solving underlying issue in #6670.

**Comment by cortinico** [comment]
> So you cannot exclude whole source sets that way or even detect which source set is in `build` directory

Can't you just disable the Gralde task for the source set you don't want to inspect?

**Comment by matejdro** [comment]
As far as I understand, no:

* `detekt` task will just scan all the source sets inside a module
* `detektMain` / `detektDebug` / `detektTest` will scan all source sets matching that flavor. And there are `main` source sets in both `src` and `build/generated`

So there is nothing to disable, current detekt tasks will always include generated files

**Comment by cortinico** [comment]
> * `detektMain` / `detektDebug` / `detektTest` will scan all source sets matching that flavor. And there are `main` source sets in both `src` and `build/generated`

That's not the case.

https://github.com/detekt/detekt/blob/df41ed7944ca25495aa4d257db0f45f6305223e0/detekt-gradle-plugin/src/main/kotlin/dev/detekt/gradle/plugin/internal/DetektAndroidCompilations.kt#L23-L72

`detektMain` will just invoke other tasks that you can selectively exclude.

What is the output of your `./gradlew detektMain --dry-run` ?

**Comment by matejdro** [comment]
Yes, there are other tasks being invoked, but they are linked to the variants, not to the source sets.

With `detektMain --dry-run`, it will run these tasks:

```
:app:detektDebug SKIPPED
:app:detektRelease SKIPPED
:app:detektMain SKIPPED
```

But, for example, `detektDebug` task (which, as far as I can tell, is not subdivided into individual tasks anymore) contains following source sets:

* `build/generated/ksp/debug`
* `build/generated/source/kaptDebug`
* `build/generated/source/kaptDebug`
* `src/main`
* `src/debug`

note that there are source sets for debug variant in both generated and src folders.

**Comment by cortinico** [comment]
Ah yes I see what's happening. I recall we had this problem already in the past.
@3flex do you recall what's the best way forward here?

**Comment by matejdro** [comment]
This is still relevant

**Comment by SteliosPapamichail** [comment]
Any updates/workarounds on this?

**Comment by matejdro** [comment]
Workaround is this: https://github.com/detekt/detekt/issues/5611#issuecomment-1364313507

**Comment by SteliosPapamichail** [comment]
@matejdro Running `./gradlew detekt` from the root of the project locally seems to correctly ignore the /build dir using a combination of the above approaches. However, when running on a GitLab pipeline that runs on our own mac-mini, the build dir does get included in the scans. Have you encountered such an issue before?

**Comment by matejdro** [comment]
Sorry,no idea. Maybe your build pipeline calls a different task?

**Comment by SteliosPapamichail** [comment]
It calls the same task sadly but via:

```yaml
lint:
  extends: .lint
  stage: code quality
  tags:
    - macos
  interruptible: true
  allow_failure: true
  script:
    - ./gradlew detekt
  artifacts:
    when: always
    reports:
      codequality: build/reports/detekt/gitlab.json
  rules:
    - if: $CODE_QUALITY_DISABLED
      when: never
    - if: $CI_COMMIT_BRANCH```

**Comment by BraisGabin** [comment]
Related with #6612. That one was closed because it was an issue on the KGP. If there is still no solution for it we need to implement a workaround for sure.

I also added a comment here: https://youtrack.jetbrains.com/issue/KT-45161/Gradle-Support-registering-generated-sources-with-the-Kotlin-model

**Comment by 3flex** [comment]
This is now an issue for detekt itself thanks to https://github.com/detekt/detekt/pull/8563

There are two options I see in the meantime until this is fixed properly in KGP:
1. Set a default ignore on `build/generated` or `build` in the Gradle plugin
2. On YouTrack someone suggested checking task outputs from tasks that generate code and removing those from the set of input files (this was related to KSP which has issues here too). That might be a solution but in the case of the buildconfig plugin I'm not sure if there actually is a task because the files are generated during project sync, not as a result of running a task.

**Comment by cortinico** [comment]
> * Set a default ignore on `build/generated` or `build` in the Gradle plugin

This sounds like a sensible default to me. The `build` folder can be safely excluded as is quite a standard name build tools use to store build artifacts and generated code.

**Comment by 3flex** [comment]
Unfortunately I've discovered that it's not that simple.

Source directories have a root path which is not the project path. For normal source sets the root is something like `src/main/kotlin` (or `src/main`, on phone and can't confirm).

Generated sources that configure source sets properly will have something like `build/generated/buildConfig/kotlin` as its root. So applying an `exclude` with `build/` won't work because build dir isn't part of the source path, it's part of the source path root directory.

DGP uses the KGP configured source paths as the default paths to analyse when analysing a source set or compilation.

That's why when the build config plugin was applied to detekt and I had to exclude the generated source files from analysis the exclude path wasn't simply set to build or some arbitrary path.

There will be ways around this I'm sure but it won't be simple. It might be better to document instructions to properly configure exclusions.

**Comment by matejdro** [comment]
Did the method I outlined in the first post not work?

**Comment by G00fY2** [comment]
> Workaround is this: [#5611 (comment)](https://github.com/detekt/detekt/issues/5611#issuecomment-1364313507)

We run into issues after adding this workaround. Gradle (9.0) was failing because of errors when storing the configuration cache.
```
cannot deserialize object of type `org.gradle.api.Project` as these are not supported with the configuration cache.
    - field `$this_applyDetekt` of `com.example.app.configurations.DetektConfigurationKt$applyDetekt$2$2`
        - bean of type `com.example.app.configurations.DetektConfigurationKt$applyDetekt$2$2`
            - field `patternSet` of `dev.detekt.gradle.Detekt`
```

We ended up with a simpler version that currently fits our needs:

```Kotlin
    tasks.withType<Detekt>().configureEach {
       ...
       exclude {
            it.file.absolutePath.contains("/build/generated/")
        }
    }
```

**Updated version:**
```Kotlin
    val projectRootDir = layout.projectDirectory.asFile
    val projectBuildDir =  layout.buildDirectory.get().asFile

    exclude {
        it.file.relativeTo(projectRootDir).startsWith(projectBuildDir.relativeTo(projectRootDir))
    }
```

**Comment by 3flex** [comment]
Sorry I was too focused on recent comments and didn't look at the whole history and suggested workarounds.

Maybe what we should do in the plugin is to have a single FileTree for sources that has the project path as the root. Then all include/exclude patterns will be relative to the project root dir instead of source dirs. This would be inconsistent with other Gradle tasks but would be consistent with how exclusions in detekt config files are applied.

**Comment by matejdro** [comment]
It's not in the history, it's the first post / description of this issue 😄 

Having a single fileTree would be slower (since detekt Gradle task would also have to scan irrelevant folders) and it might be a problem with variant gradle tasks (for example `detektDebug` should only scan `/src/debug`. If we place the source on root, we have to filter that manually).

**Comment by BraisGabin** [comment]
#8933 should close this, right? They are the same issue.

**Comment by matejdro** [comment]
They are similar, but not the same. #8933 is about automatically excluding generated sources, while this is about giving user an option to select which sources to exclude.

**Comment by BraisGabin** [comment]
Maybe i miss read the issue. But is there other good use case for that than the generated one?

**Comment by matejdro** [comment]
Yeah, the issue is 90% about generated. Maybe some would want to exclude tests or other non-main source sets?

Generating plugin may also not properly comply and mark the sources as generated (especially now, until everyone adopts this), requiring manual intervention.
